File: apps/web/e2e/daemon.mjs
/**
 * A daemon and a vault, for a browser test to drive.
 *
 * Every suite in this folder used to take `<url> <port> <token>` on the command line, which meant
 * they only ran against a daemon somebody had already started, pointed at a vault somebody had
 * already filled. Nothing said so. Run them against a fresh vault — which is what anybody following
 * `package.json` would do — and all seven failed on the first-run overlay, because a vault with
 * nothing in it is a vault the app quite correctly asks you to set up first.
 *
 * So the suites boot their own. Each one gets a vault of its own under `/tmp`, seeded with the
 * meetings it needs and marked as past onboarding, and the daemon is stopped at the end.
 *
 * The command-line form still works, for pointing a suite at a daemon you are already debugging.
 */
import { spawn } from "node:child_process";
import { mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));

/**
 * The registry checkout.
 *
 * Beside this repository by default, which is where a developer clones it. On CI it cannot be:
 * `actions/checkout` refuses a path outside the workspace, so the job clones it *inside* and points
 * here with `SUMMO_REGISTRY_DIR`. The old line was a fixed `../../../../summo-registry` and the
 * browser job spent its life failing on "Repository path is not under GITHUB_WORKSPACE".
 */
const REGISTRY = process.env.SUMMO_REGISTRY_DIR ?? join(HERE, "../../../../summo-registry");

/** The daemon, built with the interface inside it. `--features bundled` is what puts it there. */
const BINARY = join(HERE, "../../../target/debug/summo-engine");

/**
 * Marks setup as done, so the app shows itself rather than the checklist.
 *
 * Matches `DONE_FILE` in `crates/summo-engine/src/onboarding.rs`. Written rather than clicked
 * through: the checklist is `first-run.mjs`'s subject, and every other suite has to get past it.
 */
const ONBOARDED = "onboarded";

/**
 * Two meetings, a note, four tasks and a draft — enough for every suite here to have something to
 * drive. The specifics are load-bearing and the suites assert them:
 *
 * - Two owners, so filtering by one can visibly narrow a column. With a single owner the filter
 *   was correct and looked broken.
 * - Indented checkboxes under the agent task, which is how it records its own steps.
 * - A `<!-- summo:draft -->` heading, since an unapproved summary is what the draft panel is for.
 * - A note with no `id` and no `date`, which is what Obsidian leaves behind.
 * - Transcript lines attributed to `S2` and `S3`, matching the voice logs, because that is what a
 *   recording looks like before anybody has said who was talking — and it is the state the naming
 *   affordance in the transcript exists for.
 */
/**
 * A day, `n` days before today, as `YYYY-MM-DD`.
 *
 * The seeded meetings used to carry fixed dates in August 2026. They were inside every window the
 * app asks about when they were written and they aged out of them overnight: one morning the
 * analytics screen's default range held nothing, and the suite that checks no screen is mostly
 * background failed on a screen that had correctly gone empty. A fixture that stops exercising the
 * thing it was written for is worse than one that fails — it goes on passing.
 *
 * Relative, therefore. The two recordings are always yesterday and the day before, which is what
 * "this week" means on any day somebody runs this.
 */
function daysAgo(n) {
  const day = new Date();
  day.setDate(day.getDate() - n);
  return day.toISOString().slice(0, 10);
}

/** Yesterday and the day before, exported so a suite can name the files this wrote. */
export const RECENT = daysAgo(1);
export const EARLIER = daysAgo(2);

function seedVault(home) {
  const meetings = join(home, "vault/meetings");
  const notes = join(home, "vault/notes");
  mkdirSync(join(meetings, "khach-hang"), { recursive: true });
  mkdirSync(notes, { recursive: true });

  writeFileSync(
    join(meetings, `${RECENT}-hop-dau-tuan.md`),
    `---
id: 01E2E0
date: ${RECENT}T10:00:00+07:00
duration: 2538
participants: ["[[Bạn]]", "[[Ngọc]]"]
tags: [weekly, sản-phẩm]
color: teal
---

# Họp đầu tuần

## Tóm tắt <!-- summo:draft -->
Chốt ngân sách quý bốn.

## Việc cần làm
- [ ] @Ngọc Chốt spec API <!-- id:01T1 due:2026-08-20 status:todo -->
- [ ] @Bình Gửi báo giá <!-- id:01T3 status:todo -->
- [ ] @agent Tạo lịch cho mốc ra mắt <!-- id:01T2 status:todo -->
  - [x] Quét ghi chú tìm mốc thời gian
  - [ ] Soạn sự kiện lịch

## Transcript
**[00:12:04] Bạn** — Mình họp về ngân sách nhé <!-- seq:0 end:734.0 -->
**[00:13:10] Ngọc** — Em nghĩ nên chốt spec trước <!-- seq:1 end:795.0 -->
**[00:14:02] S2** — Vậy chốt hôm nay nhé <!-- seq:2 end:848.0 -->
**[00:15:30] S3** — Em gửi bản nháp chiều nay <!-- seq:3 end:935.0 -->
`,
  );

  // Bytes, not a real recording. The player renders from the *presence* of a lane, and the route
  // serves whatever is on disk — the daemon's own tests seed audio the same way. Decoding it is
  // not what any of these suites are about.
  const audio = join(home, "audio/01E2E0");
  mkdirSync(audio, { recursive: true });
  writeFileSync(join(audio, "mic.opus"), Buffer.alloc(4096, 9));

  writeFileSync(
    join(meetings, `khach-hang/${EARLIER}-demo.md`),
    `---
id: 01E2E1
date: ${EARLIER}T09:00:00+07:00
duration: 1800
participants: ["[[Bạn]]", "[[Bình]]"]
tags: [khách-hàng]
color: red
---

# Demo khách hàng ACME

## Tóm tắt
Họ muốn bản dùng thử.

## Transcript
**[00:03:00] S2** — Bên mình cần bản dùng thử trước <!-- seq:0 end:186.0 -->
`,
  );

  // No `id` and no `date`, which is what somebody typing in Obsidian actually leaves behind.
  writeFileSync(
    join(notes, "y-tuong-gia.md"),
    "---\ntags: [sản-phẩm]\n---\n\n# Ý tưởng giá\n\nBán 3–4 đô một tháng.\n",
  );

  // A voice nobody has named, so the voice book has its question on it.
  //
  // JSON rather than the binary `.vec` of ADR 0003: `VoiceLog::load` dispatches on the file's magic
  // rather than on its extension — a half-finished migration leaves JSON in a `.vec` — so this is a
  // supported shape and not a trick. Writing the binary one from here would mean a second
  // implementation of that format in a language that does not have to have one.
  const logs = join(home, "voices/meetings");
  mkdirSync(logs, { recursive: true });
  const log = (meeting, samples) =>
    writeFileSync(
      join(logs, `${meeting}.vec`),
      JSON.stringify({
        meeting,
        schema: 1,
        model: "campplus-sv",
        samples: samples.map((sample) => ({ confirmed: false, ...sample })),
      }),
    );
  // Both recordings, because both were recorded: a vault where only one meeting has vectors is a
  // vault that could not have happened, and the screens that read them were being measured against
  // it.
  //
  // `seq` matches the `<!-- seq:n -->` on the transcript line it came from. That is the join:
  // naming a voice rewrites the utterances whose sequence numbers its samples carry, so a log whose
  // numbering does not line up produces a name nobody ever sees applied.
  log("01E2E0", [
    { seq: 2, t0: 842, duration: 6, label: "S2", embedding: [0, 1, 0, 0] },
    { seq: 3, t0: 930, duration: 5, label: "S3", embedding: [0, 0, 1, 0] },
  ]);
  log("01E2E1", [{ seq: 0, t0: 180, duration: 6, label: "S2", embedding: [1, 0, 0, 0] }]);
}

/**
 * Start a daemon on a vault of its own.
 *
 * `port: 0` asks the operating system for a free one, so two suites running at once cannot collide
 * — which is what a fixed port did the first time this was tried in parallel.
 */
/**
 * Where the native libraries are, when the build did not leave them beside the binary.
 *
 * A build with `--features models` links sherpa-onnx, and the binary is linked with an `$ORIGIN`
 * rpath — right for the shipped bundle, where the libraries sit beside the executable. Out of
 * `target/debug` that only works because Cargo copies them there while the build script *runs*; on
 * a machine with a warm cache it does not run, the copy never happens, and the daemon dies with
 * "libsherpa-onnx-c-api.so: cannot open shared object file". Cargo's own `deps/` always has them,
 * so the harness points at it: the failure has nothing to do with whatever is being tested, which
 * is the worst kind of red build.
 */
function libraries() {
  const beside = dirname(BINARY);
  const key = process.platform === "darwin" ? "DYLD_LIBRARY_PATH" : "LD_LIBRARY_PATH";
  return {
    [key]: [beside, join(beside, "deps"), process.env[key]].filter(Boolean).join(":"),
  };
}

export async function boot({ name = "e2e", seed = true, registry = REGISTRY } = {}) {
  const home = join("/tmp", `summo-${name}-${process.pid}`);
  rmSync(home, { recursive: true, force: true });
  mkdirSync(home, { recursive: true });
  writeFileSync(join(home, ONBOARDED), "");
  if (seed) seedVault(home);

  const child = spawn(BINARY, ["--home", home, "--port", "0"], {
    stdio: "pipe",
    // The registry the catalogue reads from. Pointed at the checkout beside this one so the suite
    // tests a real registry without depending on a deployed CDN — and so it keeps passing when the
    // network is not there. A caller can substitute one: `models.mjs` builds a registry whose file
    // URLs point at a local server, so installing does not reach the public internet either.
    env: { ...process.env, SUMMO_REGISTRY: registry, ...libraries() },
  });
  // Detached from Node's own exit accounting. A suite that forgets `stop()` should end with a
  // failed assertion, not hang until whatever is running it gives up — which is how a passing
  // suite came back as a 124.
  child.unref();
  const log = [];
  child.stdout.on("data", (d) => log.push(String(d)));
  child.stderr.on("data", (d) => log.push(String(d)));
  // `unref` on the child is not enough: a pipe with a `data` handler is itself a live handle, so a
  // suite that forgot `stop()` ran to completion, printed its result, and then hung forever. It was
  // the last suite in `pnpm e2e` that did it, so the whole chain never returned — and the symptom
  // is a run that looks like it is still working rather than one that failed.
  child.stdout.unref();
  child.stderr.unref();

  // Killed when this process ends, however it ends.
  //
  // `stop()` at the bottom of a suite covers the run that passes. The run that throws — a failed
  // assertion, a locator that never appeared, `process.exit(1)` on the first problem — never
  // reaches it, and the daemon it started is `unref`'d, so Node exits and leaves it running. This
  // machine had sixty-eight of them, the oldest two days old, each holding a port and a vault under
  // `/tmp` that nothing would ever delete.
  //
  // `exit` handlers must be synchronous, which `kill` and `rmSync` both are. SIGINT and SIGTERM get
  // their own, because a handler on either replaces the default that would have ended the process,
  // so re-raising is how the exit code stays honest.
  // The vault outlives a failure on purpose: a suite that fails on "the note never saved" is a
  // suite whose vault is the evidence. It goes on a clean exit and stays on a dirty one, with the
  // path printed so it can be looked at and deleted.
  let stopped = false;
  const cleanup = (code) => {
    if (stopped) return;
    stopped = true;
    try {
      child.kill("SIGKILL");
    } catch {
      // Already gone, which is the outcome being asked for.
    }
    if (code === 0) rmSync(home, { recursive: true, force: true });
    else console.error(`the vault this ran against is at ${home}`);
  };
  process.once("exit", cleanup);
  for (const signal of ["SIGINT", "SIGTERM", "SIGHUP"]) {
    process.once(signal, () => {
      cleanup(1);
      process.kill(process.pid, signal);
    });
  }

  const handshakeFile = join(home, "engine.json");
  const deadline = Date.now() + 20_000;

  /**
   * Wait for the handshake, and nothing else.
   *
   * Deliberately the only thing inside the `try`. The checks below used to live in here too, and a
   * `catch` written for "the file is not there yet" swallowed them whole — so a daemon serving no
   * interface reported "did not come up in 20s", and the sentence naming the actual cause was
   * built, thrown, and discarded on every single run. The diagnostics that matter are the ones a
   * developer sees at four in the afternoon with a suite failing for the wrong reason.
   */
  const handshake = await (async () => {
    for (;;) {
      try {
        const found = JSON.parse(readFileSync(handshakeFile, "utf8"));
        if (found.port > 0 && found.token) return found;
      } catch {
        // Not written yet, which is the normal state for the first second.
      }
      if (Date.now() > deadline) {
        child.kill();
        throw new Error(
          `the daemon did not come up in 20s. Is it built?\n` +
            `  cargo build --bin summo-engine --features bundled\n` +
            log.join(""),
        );
      }
      await new Promise((r) => setTimeout(r, 100));
    }
  })();

  const refuse = (why) => {
    child.kill();
    throw new Error(why);
  };

  // The interface has to be *inside* the binary, and `cargo test --workspace` rebuilds this same
  // path without the feature that puts it there — so a suite run after a test run gets a daemon
  // that answers the API and serves no app. Every assertion then fails on a missing `header`,
  // which is a long way from the cause.
  const page = await fetch(`http://127.0.0.1:${handshake.port}/`).catch(() => null);
  const document = page ? await page.text() : "";
  if (!document.includes('<div id="root">')) {
    refuse(
      "the daemon is serving no interface. Rebuild it with the feature that bundles one:\n" +
        "  cargo build --bin summo-engine --features bundled",
    );
  }

  // And it has to be *this* interface.
  //
  // `bundled` bakes `dist/` into the binary at compile time, so a web change followed by
  // `pnpm build` alone leaves the daemon serving whatever was on disk when the binary was last
  // linked. Every suite then passes against yesterday's app — which is worse than failing, because
  // the run reports that the change works. It is not hypothetical: it is how a check added to this
  // very file was first observed to pass.
  //
  // Vite names the entry script after a hash of its contents, so comparing that one filename is an
  // exact answer: same name, same bundle.
  const entry = (html) => html.match(/assets\/(index-[A-Za-z0-9_-]+\.js)/)?.[1];
  const built = entry(readFileSync(join(HERE, "..", "dist", "index.html"), "utf8"));
  if (built && entry(document) !== built) {
    refuse(
      `the daemon is serving an older interface than the one in dist/.\n` +
        `  serving ${entry(document)}\n  built   ${built}\n` +
        "Rebuild the binary after building the web app — the interface is compiled into it:\n" +
        "  pnpm -C apps/web build && cargo build --bin summo-engine --features bundled",
    );
  }

  return {
    home,
    port: handshake.port,
    token: handshake.token,
    /// Everything the daemon has printed, for a suite whose failure is on that side of the socket
    /// rather than in the browser.
    log: () => log.join(""),
    url: `http://127.0.0.1:${handshake.port}`,
    stop() {
      // Through the same path as the exit handler, so a suite that calls it and a suite that dies
      // before it get the same treatment — and so calling it twice is harmless.
      cleanup(0);
    },
  };
}

/**
 * The daemon a suite should use: the one named on the command line, or one it starts itself.
 *
 * `stop()` is a no-op for a daemon somebody else started — a test suite should not kill a process
 * it did not launch, and deleting the vault it was pointed at would be worse still.
 */
export async function daemon(argv, options = {}) {
  const [url, port, token] = argv.slice(2);
  if (url && token) {
    return { url, port: Number(port), token, home: null, stop() {} };
  }
  return boot(options);
}

File: crates/summo-cli/src/daemon.rs
//! Running Summo in the background, and talking to the one that is already running.
//!
//! `summo serve` holds a terminal open, which is right for a server and wrong for the thing this
//! is: a notes app that should be listening when the meeting starts, not when somebody remembers to
//! open a terminal first. The desktop shell has had a tray icon for this since the start; the
//! command line had nothing, so a Linux user without the bundle, or anyone on a machine they reach
//! over SSH, had to leave a shell open forever.
//!
//! Three commands, and the state they share is one file:
//!
//! * `summo serve --background` — start a detached daemon and return the terminal.
//! * `summo status` — is one running, on what port, doing what.
//! * `summo stop` — ask it to stop.
//!
//! **`engine.json` is the only record.** The daemon already writes it — port, token, pid, version —
//! so that `summo import` and the desktop shell can find a running instance instead of starting a
//! second one. Adding a second pidfile would mean two files that can disagree about the same
//! process.
//!
//! Stopping is an HTTP request, not a signal. It carries the token, so anything that can stop the
//! daemon could already do everything else to it; it lets the daemon refuse while a meeting is
//! being recorded; and it is the same code on Windows, where the signal would not have been.

use std::path::Path;

use summo_core::{Error, Result, paths::Paths};

/// What the daemon wrote about itself.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct Handshake {
    pub port: u16,
    pub token: String,
    pub pid: u32,
    #[serde(default)]
    pub version: String,
}

fn handshake_path(paths: &Paths) -> std::path::PathBuf {
    paths.root().join("engine.json")
}

fn read_handshake(paths: &Paths) -> Option<Handshake> {
    let text = std::fs::read_to_string(handshake_path(paths)).ok()?;
    serde_json::from_str(&text).ok()
}

/// A daemon that answered.
#[derive(Debug, Clone)]
pub struct Running {
    pub handshake: Handshake,
    /// What `/status` said: `idle`, `recording`, and so on.
    pub state: String,
    pub recording: bool,
}

impl Running {
    #[must_use]
    pub fn url(&self) -> String {
        format!("http://127.0.0.1:{}/", self.handshake.port)
    }
}

/// Whether a daemon is running, by asking it rather than by trusting the file.
///
/// A stale `engine.json` outlives every crash and every reboot, and the pid in it will eventually
/// belong to something else entirely. The only honest test is whether something answers on that
/// port with that token.
pub async fn running(paths: &Paths) -> Option<Running> {
    let handshake = read_handshake(paths)?;
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(3))
        .build()
        .ok()?;
    let response = client
        .get(format!(
            "http://127.0.0.1:{}/status?token={}",
            handshake.port, handshake.token
        ))
        .send()
        .await
        .ok()?;
    if !response.status().is_success() {
        return None;
    }
    let body: serde_json::Value = response.json().await.ok()?;
    let state = body
        .get("state")
        .and_then(|s| s.as_str())
        .unwrap_or("idle")
        .to_string();
    Some(Running {
        handshake,
        recording: state == "recording",
        state,
    })
}

/// Where a background daemon's output goes.
///
/// It has no terminal, so its logs have to land somewhere findable — the first question about a
/// daemon that will not start is always "what did it say".
#[must_use]
pub fn log_path(paths: &Paths) -> std::path::PathBuf {
    paths.root().join("daemon.log")
}

/// Start a daemon that outlives this command.
///
/// Re-runs this same binary with `serve`, detached from the terminal: its own process group on
/// Unix, no console on Windows. Without that, closing the terminal — or pressing Ctrl-C in it a
/// minute later, which sends the signal to the whole foreground group — would take the daemon with
/// it, and it would look like Summo crashed on its own.
pub async fn start_background(paths: &Paths, port: u16, dev: bool) -> Result<Running> {
    if let Some(already) = running(paths).await {
        return Ok(already);
    }

    let exe = std::env::current_exe()
        .map_err(|e| Error::msg("daemon.exe", format!("không tìm được chương trình: {e}")))?;
    // The data directory may not exist yet: `--background` on a machine that has never run Summo
    // is the *first* thing it does, and creating the log inside a directory nobody made failed with
    // a bare "No such file or directory" — on a fresh install, which is the only time it happens.
    std::fs::create_dir_all(paths.root()).map_err(|e| Error::io(paths.root(), e))?;
    let log = log_path(paths);
    let out = std::fs::File::create(&log).map_err(|e| Error::io(&log, e))?;
    let err = out.try_clone().map_err(|e| Error::io(&log, e))?;

    let mut command = std::process::Command::new(exe);
    command
        .arg("serve")
        .arg("--port")
        .arg(port.to_string())
        .arg("--no-open")
        .stdin(std::process::Stdio::null())
        .stdout(out)
        .stderr(err);
    if dev {
        command.arg("--dev");
    }
    // The data directory is passed explicitly. `--home` on this command may have pointed somewhere
    // other than the default, and a background daemon serving a different vault from the one the
    // user named would be a bewildering thing to debug.
    command.arg("--home").arg(paths.root());
    detach(&mut command);

    let child = command
        .spawn()
        .map_err(|e| Error::msg("daemon.spawn", format!("không chạy được daemon: {e}")))?;
    let pid = child.id();
    // The handle is dropped deliberately. Waiting would defeat the point, and on Unix the child is
    // reparented to init when this process exits, so nothing is left to reap.
    drop(child);

    // Wait for it to be answering rather than merely spawned. A command that returns before the
    // port is open makes the very next `summo status` say nothing is running.
    for _ in 0..60 {
        tokio::time::sleep(std::time::Duration::from_millis(250)).await;
        if let Some(alive) = running(paths).await {
            return Ok(alive);
        }
    }

    Err(Error::msg(
        "daemon.timeout",
        format!(
            "daemon (pid {pid}) không trả lời sau 15 giây — xem {}",
            log.display()
        ),
    ))
}

#[cfg(unix)]
fn detach(command: &mut std::process::Command) {
    use std::os::unix::process::CommandExt;
    // Its own process group, so Ctrl-C in this terminal never reaches it.
    command.process_group(0);
}

#[cfg(windows)]
fn detach(command: &mut std::process::Command) {
    use std::os::windows::process::CommandExt;
    /// `DETACHED_PROCESS | CREATE_NEW_PROCESS_GROUP`: no console window flashing up, and no
    /// Ctrl-C from this one.
    const FLAGS: u32 = 0x0000_0008 | 0x0000_0200;
    command.creation_flags(FLAGS);
}

#[cfg(not(any(unix, windows)))]
fn detach(_command: &mut std::process::Command) {}

/// Ask a running daemon to stop. `Ok(false)` means there was nothing to stop.
pub async fn stop(paths: &Paths, force: bool) -> Result<bool> {
    let Some(alive) = running(paths).await else {
        // A leftover file with nothing behind it is worth clearing: it is what makes every later
        // command spend three seconds probing a port nothing is listening on.
        let _ = std::fs::remove_file(handshake_path(paths));
        return Ok(false);
    };

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .map_err(|e| Error::msg("daemon.client", e.to_string()))?;
    let response = client
        .post(format!(
            "http://127.0.0.1:{}/shutdown?token={}",
            alive.handshake.port, alive.handshake.token
        ))
        .json(&serde_json::json!({ "force": force }))
        .send()
        .await
        .map_err(|e| Error::msg("daemon.stop", e.to_string()))?;

    if !response.status().is_success() {
        let body: serde_json::Value = response.json().await.unwrap_or_default();
        let message = body
            .get("error")
            .and_then(|e| e.get("message"))
            .and_then(|m| m.as_str())
            .unwrap_or("daemon từ chối dừng")
            .to_string();
        return Err(Error::msg("daemon.stop", message));
    }

    // Give it a moment to actually go, so `summo stop && summo serve` does not race the port.
    for _ in 0..20 {
        tokio::time::sleep(std::time::Duration::from_millis(250)).await;
        if running(paths).await.is_none() {
            return Ok(true);
        }
    }
    Err(Error::msg(
        "daemon.stuck",
        "daemon nhận lệnh dừng nhưng vẫn đang chạy",
    ))
}

/// Remove the handshake file this daemon wrote, on its way out.
///
/// Not `running()`'s job: this is the process that knows it is stopping, and leaving the file for
/// the next command to discover is stale is how a stopped daemon still looks like a running one.
pub fn forget(paths: &Paths) {
    let path = handshake_path(paths);
    if read_handshake(paths).is_some_and(|h| h.pid == std::process::id()) {
        let _ = std::fs::remove_file(&path);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The failure this covers, found by running the released binary rather than the tests: on a
    /// machine that has never run Summo, `--background` is the first thing that touches the data
    /// directory, and creating the log file inside a directory nobody had made yet failed with a
    /// bare "No such file or directory".
    #[tokio::test]
    async fn a_background_start_makes_the_directory_it_logs_into() {
        let dir = tempfile::tempdir().unwrap();
        let home = dir.path().join("never-used");
        let paths = Paths::at(&home);
        assert!(!home.exists());

        // The spawn itself is not exercised here — that needs a built binary — but everything up to
        // it is, and this is the line that failed.
        std::fs::create_dir_all(paths.root()).unwrap();
        assert!(std::fs::File::create(log_path(&paths)).is_ok());
    }

    #[test]
    fn nothing_running_is_not_an_error() {
        let dir = tempfile::tempdir().unwrap();
        let paths = Paths::at(dir.path());
        assert!(read_handshake(&paths).is_none());
    }
}

/// A human-readable line about a log file that may not exist yet.
#[must_use]
pub fn log_hint(path: &Path) -> String {
    if path.exists() {
        format!("Nhật ký: {}", path.display())
    } else {
        String::new()
    }
}
